### PR TITLE
Implement a DisplayList mechanism similar to the Skia SkLiteDL mechanism

### PR DESCRIPTION
**Pull request by flar**
The mechanism is a drop-in replacement for:
```
  SkPictureRecorder -> DisplayListBuilder
  SkPicture         -> DisplayList
  SkCanvas          -> DisplayListCanvasDispatcher
```

The current Flutter Picture/Canvas mechanism redirects (will redirect) to DisplayList with minimal changes

Fixes: https://github.com/flutter/flutter/issues/53501
Replaces: https://github.com/flutter/engine/pull/25234

This is a new approach to implementing a replacement for SkPicture than the previous attempt (https://github.com/flutter/engine/pull/25234). In this approach I used a native DisplayList mechanism heavily based on Skia's SkLiteDL with some tweaks for our needs. Unlike 25234 which required extensive modification of the Dart Picture, PictureRecorder, and Canvas classes, this implementation will slide in natively by mimicking the SkCanvas interface and will be used directly by the native counterpart to Canvas. Thus, the dart:ui classes are not aware that they are using a new storage format.

Some of the tests written for this mechanism turned up bugs in Skia which are actively being worked on. ~At least one is likely already integrated so I can remove a workaround as soon as I can verify and test with it.~ (fix integrated and workaround removed for ColorFilter)

Some caveats on this WIP:
- It is 100% functional and runs all of the apps I've tested with
- Most of the code is in new files in `flow/display_list*` and `flow/layers/display_list_layer*`
- It is currently enabled by default to run the pre-integration tests, but due to the need for stability will be turned off by default (opt-in) before merging
- The opt-out/opt-in is controlled using `--dart-flag=--enable-display-list` or `--dart-flag=--no-enable-display-list`
- Testing should be complete now ~is nearly complete. Most all of the functionality needed by Flutter is tested (Text is a notable exception) and~ including some of the cases that are only needed to mimic SkCanvas fully are tested as well.
- Every rendering primitive ~(\~90%)~ should be tested against every rendering attribute ~(\~99%)~ in `display_list_canvas_unittests.cc`

### CODE REVIEW

File: flow/display_list.cc
Comment (dnfield) [inline flow/display_list.cc:714]: Commented out code.

File: flow/display_list.cc
Comment (dnfield) [inline flow/display_list.cc:759]: General question here:

Why are we using raw basically untyped pointers and pointer arithmetic? 

Are we really saving enough by avoiding virtual dispatch to make it worth having this surface for memory corruption/lack of type checking?

File: flow/display_list.h
Comment (dnfield) [inline flow/display_list.h:24]: I'd probably omit reference to SkPicture here, or maybe put it further down.  This paragraph is also a bit verbose. Maybe something just like:

```
The Flutter DisplayList encapsulates a persistent sequence of rendering operations.
```

And then maybe a paragraph at the end with references to implementations of `Dispatcher` and Skia's codebase.

File: flow/display_list.h
Comment (dnfield) [inline flow/display_list.h:140]: Why are we extending `SkRefCnt`, as opposed to using `fml::RefPtr` or `std::shared_ptr`?

File: flow/display_list.h
Comment (dnfield) [inline flow/display_list.h:187]:
```suggestion
  SkRect bounds_cull_;
```

Here and elsewhere: use snake_case for variable names.

File: flow/display_list.h
Comment (dnfield) [inline flow/display_list.h:201]:
```
  static_assert(kMaxDrawPointsCount * sizeof(SkPoint) < 1 << 32);
  static constexpr int kMaxDrawPointsCount = ...;
```

File: flow/display_list.h
Comment (dnfield) [inline flow/display_list.h:295]: Add a TODO with a link to the skia bug for this

File: flow/display_list.h
Comment (dnfield) [inline flow/display_list.h:422]: We should really, really try to avoid `void*`. Related to my other comment above.

File: flow/layers/display_list_layer.cc
Comment (dnfield) [inline flow/layers/display_list_layer.cc:68]: Why 10?

Could we somehow surface in the display list interface whether it has expensive-to-compare ops in it? For example, if there's a single op that takes 100ns to compare, but 50 ops that each take 1ns...

Do we have benchmarks to say that this is a good number to go with? If not, do we have a plan (and an issue filed) to make this more reliable/performant?

File: flow/layers/layer.h
Comment (dnfield) [inline flow/layers/layer.h:267]: I know I'm late to the party here, but rather than continuing to add to these, can we just create a virtual method that gives the layer an opportunity to virtually do what it needs to do with the other layer? It looks like we're more or less doing that with the DiffContext object in some places, but that's failing to capture some significance somehow.

File: flow/raster_cache.cc
Comment (dnfield) [inline flow/raster_cache.cc:77]: I think we've discussed this before in another context, but it seems like it'd be a mistake to ever allow a display list to have an infinite cull rect.

In particular, this could mean that someone has introduced NaN into the cull rect, which is almost certainly going to be a programming error.

File: lib/ui/compositing/scene_builder.cc
Comment (dnfield) [inline lib/ui/compositing/scene_builder.cc:290]: Can we just make this a method on `Picture` instead?

e.g. `auto layer = picture->createLayer();`?

And elsewhere: can we generally avoid needing multiple callsites to be aware of whether they need to ask for the `picture()` or the `display_list()`.

**Review by dnfield — COMMENTED** [review]
My primary concern right now is around the pointer arithmetic/void* usage. Do we have some compelling evidence to show that this is buying us enough in performance gains to avoid using C++ type safety features?

File: flow/display_list.cc
Comment (flar) [reply flow/display_list.cc:39]: To be clear, this isn't "the following will only work on a 64-bit platform", but more "if we assume the worst case alignment of a 64-bit platform, then here is my take on how much space we will waste".

To be clear, on a 32-bit platform there is almost no waste at all in any of these structures.

Comment (flar) [reply flow/display_list.cc:39]: And, as I learned via the test tasks below, our Windows compiler currently assumes 16-byte alignment which just adds a ton of extra unused byte data.

File: flow/display_list.cc
Comment (flar) [reply flow/display_list.cc:714]: For a good reason documented in so many other places. I got tired of adding the description comment after a while.

Skia doesn't expose that structure (yet). There is a bug out there. They are looking into it. I cannot fully capture all SkCanvas rendering until I can support this structure. The code will eventually be used, and it must exist for completeness, but I am blocked by their header file structuring.

https://bugs.chromium.org/p/skia/issues/detail?id=12125

File: flow/display_list.cc
Comment (flar) [reply flow/display_list.cc:759]: See the SkLiteDL reference in the description. It is Skia's recommendation for how to implement a DisplayList and it is in use by both Android and Chrome in just this form.

File: flow/display_list.h
Comment (flar) [reply flow/display_list.h:422]: Boilerplate from SkLiteDL used in both Android and Chrome. What would you suggest instead?

File: flow/layers/display_list_layer.cc
Comment (flar) [reply flow/layers/display_list_layer.cc:68]: I copied that from the PictureLayer implementation for now. It can be tuned over time. We might want to have a higher threshold given that the DL.equals() method exists and is much more efficient than the SkPicture technique of serializing the stream and comparing a hash.

And for now this is based on the count of the records in the DL. We could also compute and base this on a complexity metric. Since the vast majority of the ops are compared using a bulk compare, we could base it on the number of bytes. I think @knopp probably has some experience with how to tune this metric. For this first pass, I should probably bump this up or eliminate the test - Matt?

File: flow/layers/layer.h
Comment (flar) [reply flow/layers/layer.h:267]: This would be for @knopp to decide and would apply to the overall implementation of the DiffContext (see https://github.com/flutter/engine/pull/21824)

File: flow/raster_cache.cc
Comment (flar) [reply flow/raster_cache.cc:77]: Another piece of code stolen from the Picture cache. All of these methods in here are duplicates of the adjacent methods that take SkPicture.

If someone introduces a non-finite op then we should probably ignore the one op that did so, but not discard the picture entirely. We can at least avoid caching that picture, though.

File: lib/ui/compositing/scene_builder.cc
Comment (flar) [reply lib/ui/compositing/scene_builder.cc:290]: I only remember 2 places that dealt with this. Here and the unittest that used to assume that all pictures were SkP. The latter shouldn't leak methods into the Picture as it is just a unittest.

File: flow/display_list.cc
Comment (flar) [reply flow/display_list.cc:37]: Done

File: flow/display_list.cc
Comment (flar) [reply flow/display_list.cc:714]: (I'm hoping that they'll release the structure before I merge this. Otherwise I'll put more reasons in or switch these to ifdefs or something.)

File: flow/display_list.cc
Comment (chinmaygarde) [inline flow/display_list.cc]: Can you document what bulk compare means?

File: flow/display_list.cc
Comment (chinmaygarde) [inline flow/display_list.cc:101]: Instead of these comments, perhaps something like `static_assert(sizeof(SetStrokeWidthOp) == sizeof<DLOp> + ...)`. Just to make the compiler check the comment for you too.

File: flow/display_list.h
Comment (flar) [reply flow/display_list.h:24]: I reworked the comment.

File: flow/display_list.cc
Comment (chinmaygarde) [inline flow/display_list.cc:453]: Full disclosure, I stopped double checking each struct definition right about here as it is just repetition of the same pattern.

Perhaps (later if needed) this entire file can be generated from a manifest? It's fairly straightforward to make a target depend on a generated TU in GN. It will also allow us to switch implementations on a whim. For instance, to see if Dan's concerns make sense on whether we should just use vtable dispatch.

Even if I am unsure if we could have just gotten away with simple vtable dispatch, I couldn't spot any obvious deficiencies with this approach.

File: flow/display_list.cc
Comment (flar) [reply flow/display_list.cc]: I also switched 2 other enums to enum classes including the main one used to identify the records.

File: flow/BUILD.gn
Comment (chinmaygarde) [inline flow/BUILD.gn:15]: Minor nit: Can we put the ops in their own header? The stuff about the display list is way more interesting that the op definitions. Will just make the code a bit easier to navigate.

**Review by chinmaygarde — COMMENTED** [review]
Haven't gone over the whole thing yet but sending the comments I have. More to follow.

File: flow/display_list.h
Comment (flar) [reply flow/display_list.h:155]: For consistency with SkPicture that it is replacing? That was probably my default reasoning. Also, every other complex structure included in the DL by reference supports SkRefCnt which might be helpful if we ever want to do some other bulk operations on the DL contents.

I found in testing that sk_sp<Foo> can be compared using a bulk compare since it doesn't actually wrap the object. Basically: `sk_sp<Foo> == &Foo` so bulk comparing across a bunch of sk_sp<> pointers is equivalent to bulk comparing across a bunch of raw pointers to the objects themselves. I'm not sure about the others, but since std::shared_ptr can work on any type then I'd assume that it is a wrapper which would require a lot more care in equals comparisons.

File: flow/display_list.h
Comment (flar) [reply flow/display_list.h:216]: It actually needs to be much smaller. At the time I was looking at the "count" parameter being an int, but we encode the size of the op in a 24-bit field, so we can only support enough points to fit into the pod data area.

File: flow/display_list.h
Comment (flar) [reply flow/display_list.h:312]: I will if they don't move the struct reference soon.

File: flow/display_list.h
Comment (dnfield) [reply flow/display_list.h:155]: SGTM. As a nitpick, might worth documenting that reasoning here, since this will be the only SkRefCnt we define.

File: flow/raster_cache.cc
Comment (dnfield) [reply flow/raster_cache.cc:77]: I think most of our Dart API surface should prevent that from happening - but if somehow it's being let through we should be letting the programmer know.

File: flow/BUILD.gn
Comment (flar) [reply flow/BUILD.gn:15]: Currently they are private to that cc file. Their definitions aren't needed anywhere else.

Basically `display_list.[h,cc]` exist only to define, pack, and unpack the structure through Dispatcher. Everything else is in a different file.

File: flow/display_list.h
Comment (dnfield) [reply flow/display_list.h:441]: Chrome uses a template here: https://source.chromium.org/chromium/chromium/src/+/main:cc/paint/paint_op_buffer.h;l=1109?q=SkLiteDL&ss=chromium%2Fchromium%2Fsrc

File: flow/display_list.cc
Comment (dnfield) [reply flow/display_list.cc:780]: (let's discuss this here: https://github.com/flutter/engine/pull/26928#discussion_r658254626)

**Comment by flar** [comment]
Perhaps we need to have a meta-discussion about the structure packing and sizes.

I wasn't really paying much attention to it until I started writing the display_list_unittests.cc and I decided to assert the sizes of the DLs being generated - mostly for completeness. It seemed like it was just verifying some info that we had an API to provide (dl->bytes()), but I suddenly realized that the sizes weren't what I thought they should be. That should have probably been obvious due to alignment of data members and such, but it was a lot different than what I was expecting. For one thing, I used to have a few ops that I packed into "4 bytes" before I realized that the code imposed 8-byte alignment and for a good reason (ptr alignment). Then I realized that some structures had some major packing issues, so I went about analyzing all of them and adjusting their field orders until they had "fairly optimum" sizes. I documented my work as I went for my own tracking and possibly to inform future maintainers.

But, in the long run, it isn't a big thing to worry about. It might be nice to track that someone doesn't add a huge data structure to every Op and explode our DL sizes, but if it packs to 20 bytes on this platform and 24 on that platform, the difference isn't critical.

On the other hand, a half hour of my diligence in rearranging a few structures saved us some memory at no run-time cost, so there are some minor benefits to it. And I discovered the Windows compiler preference for 16-byte alignments which is gratuitous for our data here, so that is another side benefit.

File: flow/display_list.h
Comment (dnfield) [reply flow/display_list.h:441]: They also still have some void pointers, but some of it is abstracted away or more contained and avoids _some_ pointer arithmetic in https://source.chromium.org/chromium/chromium/src/+/main:cc/paint/paint_op_reader.cc

Part of the challenge is that code like this is really hard to audit for security/safety purposes. And that will become more true as we extend/refactor it (unless we specifically refactor it for safety auditing purposes :).  If we can do any of that now we should.

But AFAICT, SkLiteDL is not actually used anywhere in Chrome or Android.

File: flow/display_list.cc
Comment (flar) [reply flow/display_list.cc]: I added this:
```
  // The Op is deferring comparisons to a bulk memcmp performed lazily
  // across all bulk-comparable ops. Most ops fall into this category
  // if they contain raw numeric values or have sk_sp<> pointers that
  // point to objects that we cannot compare explicitly and so we rely
  // on their identities to compare ==. (Note that two sk_sp<> that
  // point to the same object are themselves identical ==.)
  kUseBulkCompare,
```

File: flow/display_list.cc
Comment (chinmaygarde) [inline flow/display_list.cc:920]: The routines here do not follow the style guide. Presumably, this is just from Skia's SkLiteDL implementation?

Comment (flar) [reply flow/display_list.cc:920]: Yes, cut and paste.

I also have ingested a few SkAssert calls that could probably change to FML_DCHECKs.

File: flow/display_list.h
Comment (flar) [reply flow/display_list.h:441]: By "using" I meant that Skia had provided SkLiteDL as a template for them to use. They customized it to their needs, but you can see the basic bones of SkLiteDL in the structure of their code. You found the Chromium variant, but Android has their own as well - similarly customized.

File: lib/ui/compositing/scene_builder.cc
Comment (flar) [reply lib/ui/compositing/scene_builder.cc:290]: Another issue is that Picture is in lib/ui/painting so having a method to produce a flow/layer/*Layer on it would be out of place. AFAIR, code in lib/ui isn't really aware of layers (but there is some cross-talk to flow/*?).

File: flow/display_list.h
Comment (flar) [reply flow/display_list.h]: Is the new comment better?

File: flow/raster_cache.cc
Comment (flar) [reply flow/raster_cache.cc:77]: What are you suggesting? DML_CHECK? FML_LOG?

File: flow/display_list.cc
Comment (flar) [reply flow/display_list.cc]: I reworked it a bit more after that with a comment on the reason for the enum to exist in the first place. Did it explain things well enough?

File: flow/display_list.cc
Comment (flar) [reply flow/display_list.cc]: I made quite a number of changes to style in https://github.com/flutter/engine/pull/26928/commits/cd7fed2fd769dccfc6735920b35ebd41db523cd0

You might want to review that commit in isolation and see if you agree with them. Many of the lines are dup/analogous changes so much of it can be scanned.

File: flow/raster_cache.cc
Comment (dnfield) [reply flow/raster_cache.cc:77]: Log is probably good, if we can provide any details about which picture or which call did it that's ideal but I doubt we can.

File: flow/display_list.h
Comment (dnfield) [reply flow/display_list.h]: Yes, thanks

File: flow/display_list.h
Comment (dnfield) [reply flow/display_list.h]: Can we customize this to avoid some of the pointer arithmetic? :)

Comment (flar) [reply flow/display_list.h]: I'm not sure what you are suggesting? Making the structs full classes? That adds overhead of a vtable per op even though the actions are only needed in one or two methods - currently those use a switch statement to inline the 2 implementations they care about (dispatch and destructor).

Since the vast majority of the entries in the list are trivially destructible the destructor can dance along the array only calling unref on the few sk_sp<> references that dot the list and then free the array. If we made them all virtual then each op would have to be manually destructed by a virtual pointer.

The comparison method can also do 99+% of its work by a memcmp in this scenario.

The memory of all of the ops is localized which would go away if we converted to some sort of virtual typesafe list of pointers to the structures. It might end up localized, but that might deteriorate over time.

Perhaps I'm not familiar with what you are suggesting, but those are the advantages that it looks like the design was aiming to accomodate.

File: flow/raster_cache.cc
Comment (flar) [reply flow/raster_cache.cc:77]: Somewhere someone was looking to add some annotation records to our "pictures". The DL could accomodate that and then it could log the last "info annotation" to provide a programmer a way to trace it back in their program.

File: flow/display_list.cc
Comment (flar) [reply flow/display_list.cc:807]: This Skia bug will be discussed in a synch-up today, so I can make a decision about this code soon.

File: flow/display_list.h
Comment (flar) [reply flow/display_list.h]: I like the suggestion from @chinmaygarde to move the op definitions into another file and format so that we can switch up the implementation more easily. We could then investigate the performance impact of something like a vector of virtual dispatch struct/classes.

The one question that I don't think could be easily answered from the above list of "advantages" of the current approach, though, would be the impact of memory fragmentation on locality and, by extension, on app performance. That situation would only come up in practice on long running apps. Has something like that shown up as an issue before?

File: lib/ui/compositing/scene_builder.cc
Comment (flar) [reply lib/ui/compositing/scene_builder.cc:290]: When SkPicture goes away here, we won't need an if or a method on Picture to manage this distinction. Since it is only 3 lines, I think I'm going to leave in the 2 pieces of code to reduce the cross-file technical debt here for the eventual elimination of SkPicture.

File: flow/display_list.cc
Comment (flar) [reply flow/display_list.cc:120]: I don't think I can static assert since the situation may end up being more compact on 32-bit compiles - which is a good thing. These comments are about an optimization to avoid poorly packed structures wasting space unnecessarily. The values are checked in a unit test that run on 64-bit hosts, so I think that is a good compromise.

File: flow/display_list.cc
Comment (flar) [reply flow/display_list.cc:515]: I like the idea for generating everything from a manifest as that would improve clarity on everything. I'll look into that as a follow-on along with investigating different storage solutions as @dnfield was suggesting.

File: flow/BUILD.gn
Comment (flar) [reply flow/BUILD.gn:15]: I'll lump this in with the previous suggestions for using a manifest and for investigating alternate storage solutions.

File: flow/display_list.cc
Comment (flar) [reply flow/display_list.cc]: I did a lot of style guide work over the past couple of days. I'm going to mark this conversation resolved and leave it for further review to mention specific issues.

File: flow/layers/display_list_layer.cc
Comment (flar) [reply flow/layers/display_list_layer.cc:68]: I switched to a byte size comparison with a guess at a threshold of 10,000 bytes. I am waiting to hear back from @knopp before I push the change.

File: flow/display_list.cc
Comment (flar) [reply flow/display_list.cc:814]: Skia says that this will take some time as they are also working on a redesign of the Shadow stuff so they would like to prioritize the Metal work over this header work. I think I'll just delete all references to the ShadowRec stuff (except, obviously, in the SkCanvas->DL adapter) for now. We only use this from Dart Canvas presently and that code already works around this issue.

File: flow/raster_cache.cc
Comment (flar) [reply flow/raster_cache.cc:78]: I added an FML_LOG(INFO) to both the Picture and DL caching methods.

File: flow/display_list.cc
Comment (flar) [reply flow/display_list.cc:123]: I beefed up the comment at the top of the section that these calculations are only "worst case" numbers to double check we aren't wasting space unnecessarily.

**Comment by flar** [comment]
My next push will remove the references to ShadowRec for now.

How do we feel about the scheduling of this PR and whether we should introduce it as an opt-in or opt-out at this time?

File: common/settings.h
Comment (chinmaygarde) [inline common/settings.h:167]: Is this on by default just for purposes of making the presubmits use the display list stuff? I think we should land this with off by default and in another patch, flip the switch.

File: shell/common/switches.cc
Comment (chinmaygarde) [inline shell/common/switches.cc:69]: These flags are meant to be sent directly to the Dart VM during initialization. For engine specific flags, we add them to `shell/common/switches.h`

File: shell/common/switches.cc
Comment (chinmaygarde) [inline shell/common/switches.cc:409]: See the comment above about adding to shell switches instead of Dart switches.

File: flow/layers/display_list_layer_unittests.cc
Comment (chinmaygarde) [inline flow/layers/display_list_layer_unittests.cc:22]: Why is this run in debug only?

**Review by chinmaygarde — APPROVED** [review]
Barring a couple of nits, I believe this patch is good to go.

While its size make it daunting to review, a substantial portion is boilerplate that we can generate from a manifest (if needed). It is conceptually sound and builds on proven work in Skia. It's also extremely well tested. There are few comment threads suggesting alternatives but I don't think there are any remaining construction issues. We have also needed this mechanism for a very long time. For these reasons, I feel comfortable in landing this right now despite the size.

The only suggestion would be to make the flag off by default with a flag that flips the switch immediately to follow. If the patch is implicated in a roll, I rather keep the delta as small as possible.

File: common/settings.h
Comment (flar) [reply common/settings.h:167]: Correct. The next commit would be to turn the flag off and remove the WIP.

File: shell/common/switches.cc
Comment (flar) [reply shell/common/switches.cc:69]: My impression is that would advertise these switches. I was looking to keep them un-advertised for now...?

File: shell/common/switches.cc
Comment (flar) [reply shell/common/switches.cc:409]: Or not - intentionally. They do work as is, but are not advertised as I don't know if we want them to become part of our public API unless we plan to support them beyond the initial testing and buy-in phase.

File: flow/layers/display_list_layer_unittests.cc
Comment (flar) [reply flow/layers/display_list_layer_unittests.cc:22]: As per all of the other layer unittests which only support the EXPECT_DEATH tests on debug? Or at least that was my interpretation of the same ifdef in all of the other unit test files. Looking around it looks like there are some similar ifdefs in the other files that wrap even non-Death tests and the final DEATH test here which is outside the ifdef as it was in the picture layer tests that I cut and pasted from.

File: common/settings.h
Comment (flar) [reply common/settings.h]: I just pushed a commit changing the flag to false by default and removed the WIP designation.

File: flow/layers/display_list_layer_unittests.cc
Comment (flar) [reply flow/layers/display_list_layer_unittests.cc:22]: I pushed a commit that moved the 2 EXPECT_DEATH tests in picture/dl_layer_unittests inside the DEBUG ifdef.

File: shell/common/switches.cc
Comment (chinmaygarde) [reply shell/common/switches.cc:69]: All switches are un-advertised and unstable really.

**Comment by chinmaygarde** [comment]
One thing to keep track of still is immediately flipping the flag to use DLs by default.

**Comment by flar** [comment]
> One thing to keep track of still is immediately flipping the flag to use DLs by default.

I actually have about 5 or 6 "follow-on" tasks to create issues for. The flag flip is waiting for successful integration down the line to set a baseline.

**Pull request merged**
